includes/abandoned-signups.php

```diff
@@ -0,0 +1,196 @@
+<?php
+
+/**
+ * Set up a user taxonomy to track users who were created during the PMPro
+ * checkout process but did not complete the checkout.
+ *
+ * @since TBD
+ */
+function pmpro_abandoned_signups_taxonomy() {
+	// Register the taxonomy.
+	register_taxonomy(
+		'pmpro_abandoned_signup',
+		'user',
+		array(
+			'public'            => false,
+			'default_term'      => array(
+				'slug' => 'abandoned-signup',
+			),
+		)
+	);
+}
+add_action( 'init', 'pmpro_abandoned_signups_taxonomy', 10 );
+
+/**
+ * Add the abandoned signup taxonomy to the user object when they are created during
+ * the PMPro checkout process.
+ *
+ * @since TBD
+ *
+ * @param int $user_id The user ID.
+ */
+function pmpro_set_abandoned_signup_taxonomy( $user_id ) {
+	// Bail if the user ID is empty.
+	if ( empty( $user_id ) ) {
+		return;
+	}
+
+	// Add the abandoned signup taxonomy to the user.
+	var_dump( wp_set_object_terms( $user_id, 'abandoned-signup', 'pmpro_abandoned_signup' ) );
+}
+add_action( 'pmpro_checkout_before_user_auth', 'pmpro_set_abandoned_signup_taxonomy', 10 );
+
+/**
+ * After a checkout is complete, remove the abandoned signup taxonomy from the user.
+ *
+ * @param int $user_id The user ID.
+ */
+function pmpro_remove_abandoned_signup_taxonomy( $user_id ) {
+	// Bail if the user ID is empty.
+	if ( empty( $user_id ) ) {
+		return;
+	}
+
+	// Remove the abandoned signup taxonomy from the user.
+	wp_remove_object_terms( $user_id, 'abandoned-signup', 'pmpro_abandoned_signup' );
+}
+add_action( 'pmpro_after_checkout', 'pmpro_remove_abandoned_signup_taxonomy' );
```

Comment:
I think we need to think of some other actions that would complete here. So e.g. something on login (but make sure to skip the login we do at checkout) would be good, because if a user logged in user that user/pass a day or two later that's an indication that they are expecting to continue using that user account.

Similarly, we need to check for posts added with them as the author (could be literal posts or maybe show that they interacted with some plugin via CPT)... or maybe specifically check for WooCommerce checkouts, EDD checkouts, etc. What other actions would indicate "this user expects to keep using this user account" besides completing a PMPro checkout?

Reply:
Let's focus on the original target issue here of spam checkouts and just folks who never come back after failing checkout. What we can do is hook into the "wp" hook and if not pmpro_is_checkout() then consider this a user who was logged in and did *something* other than try to checkout with PMPro and remove the inactive/abandoned/spam checkout taxonomy term (or whatever we end up calling it).

So we'll hook into the pmpro_after_checkout and user_deleted hooks, but add a new callback hooked on wp to check if the user has the term and remove it.

This will then run on wp for every logged in user, but should be a quick enough check that it doesn't slow things down.